### PR TITLE
refacor/CALL and OOG_CALL

### PR DESCRIPTION
**Pull request by KimiWu123**
It's for [CALL & CALL OOG refactor #940](https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/940)

### CODE REVIEW

**Comment by KimiWu123** [comment]
@ChihChengLiang , could you help to have a quick review to see if I do it in right direction? Any advice are welcome.

File: src/zkevm_specs/evm/util/call_gadget.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    callee_address: FQ
+    cd_offset: FQ
+    cd_length: FQ
+    rd_offset: FQ
+    rd_length: FQ
+
+    next_memory_size: FQ
+    memory_expansion_gas_cost: FQ
+
+    def __init__(self, instruction: Instruction, is_oog: FQ, is_call_or_callcode: FQ = FQ(0)):
```
Comment (ChihChengLiang) [inline src/zkevm_specs/evm/util/call_gadget.py:36]: Instead of a class, is it possible to make the common part of callop and oog_call a function? 

For example,

```
def call_common(instruction: Instruction):
    ...
```
It would be okay to define a struct to hold the output value.

```
class CallStruct:
    is_oog: FQ
    is_success: RLC
    value: RLC
    gas: FQ
    callee_address: FQ
    cd_offset: FQ
    cd_length: FQ
    rd_offset: FQ
    rd_length: FQ
    next_memory_size: FQ
    memory_expansion_gas_cost: FQ
```

The reason is that the Python code serves as the specification of the circuit. Organizing the constraints in the class and methods makes the reader hard to keep track of variables.

**Review by ChihChengLiang — COMMENTED** [review]
Great work. I added a suggestion on organizing the code in a function.

File: src/zkevm_specs/evm/execution/callop.py [hunk 2]
```diff
@@ -44,57 +43,37 @@ def callop(instruction: Instruction):
...
-    rd_length_rlc = instruction.stack_pop()
-    is_success = instruction.stack_push()
-
-    # Verify is_success is a bool
-    instruction.constrain_bool(is_success)
-
-    # Recomposition of random linear combination to integer
-    code_address = instruction.rlc_to_fq(code_address_rlc, N_BYTES_ACCOUNT_ADDRESS)
-
+    call_data = common_call_stack_pop(instruction, FQ(0), is_call + is_callcode)
```
Comment (ChihChengLiang) [inline src/zkevm_specs/evm/execution/callop.py:46]: The name `call_data` collides with the EVM `calldata`.
What do you think we name it `call` directly here? It would make variable names like `call.callee_address` and `call.value` look natural.

**Comment by KimiWu123** [comment]
**Update** 
After dicussed with @ChihChengLiang , we decided to refactor `zkEVM-circuit` first and then back here to see if we need more modification. The reason behind is that we would like to make the code in `zkEVM-spec` is more consistent with `zkEVM-circuit`

File: src/zkevm_specs/evm/util/call_gadget.py (hunk 1 above)
Comment (KimiWu123) [reply src/zkevm_specs/evm/util/call_gadget.py]: > The reason is that the Python code serves as the specification of the circuit. Organizing the constraints in the class and methods makes the reader hard to keep track of variables.

I changed back to class-base implementation to align with the implementation in `zkEVM-Circuit`([#1023](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1023)). Would like to know your opinion. There are some failed cases and  I'll keep working on fixing them if you're fine with this change.

Btw, I rollback @ed255's change in #354, [here ](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/354/files#diff-10487489ead1ac341b7671be200b7c7b4b63c000a33f05ffb0bcfe63e73d6ff0R131-R133). The original way is more consistent with the implementation in `zkEVM-circuit`. Do you have any concern?

File: src/zkevm_specs/evm/execution/callop.py (hunk 2 above)
Comment (KimiWu123) [reply src/zkevm_specs/evm/execution/callop.py]: fixed in [ad22495](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/348/commits/ad224959a3e90df0f79898b52888fab521c80f8b)

File: src/zkevm_specs/evm/execution/oog_call.py [hunk 3]
```diff
@@ -22,55 +15,19 @@ def oog_call(instruction: Instruction):
...
-
-    has_value = 1 - instruction.is_zero(value)
-    callee_address = instruction.rlc_to_fq(callee_address_rlc, N_BYTES_ACCOUNT_ADDRESS)
+    call = CallGadget(instruction, FQ(0), FQ(1), FQ(0), FQ(0))
 
     # TODO: handle PrecompiledContract oog cases
 
-    # Add callee to access list
-    is_warm_access = instruction.read_account_to_access_list(tx_id, callee_address)
+    # Add callee to access list``
```
Comment (ChihChengLiang) [inline src/zkevm_specs/evm/execution/oog_call.py:23]:
```suggestion
    # Add callee to access list
```

File: src/zkevm_specs/evm/util/call_gadget.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+            # Verify is_success is a bool
+            self.instruction.constrain_bool(self.is_success)
+            self.gas = instruction.rlc_to_fq(self.gas_rlc, N_BYTES_GAS)
+            self.is_u64_gas = instruction.is_zero(
+                instruction.sum(self.gas_rlc.le_bytes[N_BYTES_GAS:])
+            )
+        else:
+            self.instruction.constrain_zero(self.is_success)
+            self.gas = FQ(0)
+            self.is_u64_gas = FQ(0)
```
Comment (ChihChengLiang) [inline src/zkevm_specs/evm/util/call_gadget.py:73]: I don't think we need to constrain `gas` and `is_u64_gas` to zero. 

```suggestion
```

**Review by ChihChengLiang — COMMENTED** [review]
First pass

File: src/zkevm_specs/evm/util/call_gadget.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+    def __init__(
+        self,
+        instruction: Instruction,
+        is_success_call: FQ,
+        is_call: FQ,
+        is_callcode: FQ,
+        is_delegatecall: FQ,
+    ):
+        self.instruction = instruction
```
[the remark this answers is not in the document]
Comment (KimiWu123) [reply src/zkevm_specs/evm/util/call_gadget.py]: fixed in [53df9ee](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/348/commits/53df9ee479bd2b2f70207363595dafb265f91a96)

File: src/zkevm_specs/evm/execution/oog_call.py (hunk 3 above)
Comment (KimiWu123) [reply src/zkevm_specs/evm/execution/oog_call.py]: fixed in [53df9ee](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/348/commits/53df9ee479bd2b2f70207363595dafb265f91a96)

File: src/zkevm_specs/evm/util/call_gadget.py (hunk 4 above)
Comment (KimiWu123) [reply src/zkevm_specs/evm/util/call_gadget.py]: fixed in [53df9ee](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/348/commits/53df9ee479bd2b2f70207363595dafb265f91a96)

**Review by ChihChengLiang — APPROVED** [review]
LGTM

File: src/zkevm_specs/evm/execution/callop.py [hunk 6]
```diff
@@ -171,9 +121,9 @@ def callop(instruction: Instruction):
         is_precompile, FQ(instruction.next.execution_state in precompile_execution_states())
     )
 
-    if is_empty_code_hash == FQ(1) and is_precompile == FQ(0):
+    if call.is_empty_code_hash == FQ(1) and is_precompile == FQ(0):
```
Comment (ed255) [inline src/zkevm_specs/evm/execution/callop.py:124]: Recently the case of \*CALL\* without code execution has been updated due to how non-existing accounts are handled.
Non-existing accounts are now identified by having `code_hash = 0`, which is not the empty hash value.  But those cases should still be handled as no code execution.
So I think it would be better to define:
```python
no_callee_code = call.is_empty_code_hash + call.callee_not_exists
```
And then in this if replace the first check by `no_callee_code == FQ(1)`

I notice that `call.callee_not_exists` is missing from the `CallGadget` in this PR.  I introduced that in a very recent PR in the circuits. See https://github.com/privacy-scaling-explorations/zkevm-circuits/commit/aaaf13ee3020661e048484776bd88686d1659042#diff-08d58ebc59ab6c11bf7c049e859441eed2319a03feb3fdffaddaba73f8a8c829

File: src/zkevm_specs/evm/util/call_gadget.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        # Verify memory expansion
+        (
+            self.next_memory_size,
+            self.memory_expansion_gas_cost,
+        ) = instruction.memory_expansion_dynamic_length(
+            self.cd_offset,
+            self.cd_length,
+            self.rd_offset,
+            self.rd_length,
+        )
```
Comment (ed255) [inline src/zkevm_specs/evm/util/call_gadget.py:89]: I noticed that the fields `callee_code_hash` and `is_empty_code_hash` are not set in the `__init__`.  They are set in the `gas_cost` method.  Why is that?  Isn't it better to initialize all the Class fields in `__init__`?  The circuit does it like this;
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/b77613e9b42eacb03ccc894f0b353c374da3f283/zkevm-circuits/src/evm_circuit/util/common_gadget.rs#L523-L525

**Review by ed255 — COMMENTED** [review]
I've added a few small notes related to the latest circuit implementation, but otherwise looks good!  Please take a look at the two comments :)

File: src/zkevm_specs/evm/execution/callop.py (hunk 6 above)
Comment (KimiWu123) [reply src/zkevm_specs/evm/execution/callop.py]: fixed in [a7d329a](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/348/commits/a7d329abe49ca42ed8b9882234e04d853206f6a0)

File: src/zkevm_specs/evm/util/call_gadget.py (hunk 7 above)
Comment (KimiWu123) [reply src/zkevm_specs/evm/util/call_gadget.py:90]: yes, I totally agree that. However, I don't know how to hadle it properly because I need to get `callee_code_hash` first as following,
```
        self.callee_code_hash = instruction.account_read(
            self.callee_address, AccountFieldTag.CodeHash
        )
```

If I moved above code to `__init__`, it breaks the order of rw table. (reading the value of `AccountFieldTag.CodeHash` must be after reading the value of `AccountFieldTag.Balance`, see [here](https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/tests/evm/test_callop.py#L345-L355)). Either I change the order in the test or assign an arbitrary value in `__init__`(It will fail if anyone uses these vars before calling `gas_cost`, but I don't think it's a good idea). Do you have any better ideas for this?

Comment (ed255) [reply src/zkevm_specs/evm/util/call_gadget.py:90]: Ah I see!
After this change was applied https://github.com/privacy-scaling-explorations/zkevm-specs/pull/354 it will become impossible to read a `AccountFieldTag.Balance` for non-existing accounts.  This means that the order of lookups for possibly non-existing accounts should always be
1. Read CodeHash
2. If CodeHash != 0, then read other fields as necessary

By impossible what I mean is that it won't be possible to generate a list of MPT updates + proofs that can verify it.  Since we don't have the MPT circuit connected yet, tests will not fail (although we could add some checks to make sure we never read non-codeHash account fields of non-existing accounts).

So I think you should move the callee CodeHash lookup to happen first, and if account exists, then do the callee Balance lookup.  I think this affects the circuits as well.  I see in `error_oog_call.rs` first the callee Balance is read, and then ` call_gadget.gas_cost_expr` is called, which causes the callee CodeHash read.  Actually, I'm not sure why in `error_oog_call` the callee balance is read, there doesn't seem to be any constraint using that value: https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/ea6d8e38bcaedbf2d69d0661ad14485c6b02b9cb/zkevm-circuits/src/evm_circuit/execution/error_oog_call.rs#L66-L71

So I think you should change the order of reads in the test for the callee: first read the CodeHash, then if its != 0, read the Balance.

After a second look at the test code you mentioned https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/tests/evm/test_callop.py#L345-L355 I see that the Balance read in there is for the caller which is independent from the callee CodeHash read; which means that the order doesn't matter, but it's still OK to read the callee CodeHash before the caller Balance.

So in summary: for possibly non-existing accounts, first read codeHash, and only if its !=0 read other fields.  For accounts known to be existing in the context (like the caller in the call context), the order of reads is not important.

Comment (KimiWu123) [reply src/zkevm_specs/evm/util/call_gadget.py:90]: the order of CodeHash read was fixed in [7e0fc6e](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/348/commits/7e0fc6e7ae1880835beefa735e8fe290338eac15)
.

> Actually, I'm not sure why in error_oog_call the callee balance is read, there doesn't seem to be any constraint using that value:

yep, balance is not necessary in `error_oog_call`, I have removed it.
.

> So I think you should change the order of reads in the test for the callee: first read the CodeHash, then if its != 0, read the Balance.

For `then if its != 0, read the Balance`, I'm not sure I understand it corretly. Do you mean add a condiftion `if callee_bytecode_hash != 0` at [test_callop.py#L349](https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/tests/evm/test_callop.py#L349) and test_callop.py#L353? From my understanding, in evm, user's balance is still able to be read even it's a non-existing account (not initialized in MPT). If we add `if callee_bytecode_hash != 0`, I guess it'll panic(or just fail) if we read the balance of a non-existing account. Not sure my understanding is correct and not sure I understand your meaning correctly.

Comment (ed255) [reply src/zkevm_specs/evm/util/call_gadget.py:90]: > For `then if its != 0, read the Balance`, I'm not sure I understand it corretly. Do you mean add a condiftion `if callee_bytecode_hash != 0` at [test_callop.py#L349](https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/tests/evm/test_callop.py#L349) and test_callop.py#L353?

I was mistaken on my interpretation of the callop logic, so the `then if its != 0, read the Balance` doesn't apply here, because the Balance is read on the caller, which at this point is guaranteed to exist, so there's no need to make it conditional.  I think there's an edge case on the transfer thought, but more on it later.

> From my understanding, in evm, user's balance is still able to be read even it's a non-existing account (not initialized in MPT).

From the EVM spec, whenever an opcode that reads account fields encounters a non-existing account particular conditions are applied.  Most of the time the behavior is the same as considering the non-existing account as a default account (with nonce=0, balance=0, code_hash=EMPTY_HASH), but that's not always the case.  For example, EXTCODEHASH returns 0 for non-existing accounts (instead of EMPTY_HASH).  This also affects writes, as opcodes that write to accounts may have different gas cost when they create the account (because it was non-existing before).

In our circuits, account read/writes are triggered by the EVM circuit.  Those read/writes are then verified for consistency in the State Circuit, and finally they are collected and matched against MPT proofs to match the state trie transitions.  Our current design of the MPT Circuit supports the proofs listed in this table: https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/specs/tables.md#mpt_table

In particular, the `NonceMod`, `BalanceMod` and `CodeHashMod` don't support a read (valuePrev == value) of non-existing accounts.  They only support reads to existing accounts, and writes to existing accounts or to non-existing accounts (which triggers account creation).

So a Nonce, Balance, CodeHash read proof for non-existing accounts is not supported by the MPT circuit.

The MPT circuit supports `NonExistingAccountProof` which acts like a read and proves that an account doesn't exist.

And in the StateCircuit we map `CodeHash` read of 0 to `NonExistingAccountProof` in the MPT.  We can do this because code_hash will never be 0 in an existing account, so wenever in the EVM/State circuit a CodeHash is read as 0, it means that the account doesn't exist.  And if the account doesn't exist, a following Nonce or Balance read (triggered by the EVM) can't be proved by the MPT Circuit.

This is why every time a Nonce/Balance is read from the EVM circuit, we must make sure that the account exists, otherwise it may be translated to a  `NonceMod` or `BalanceMod` proof in the MPT that is impossible to succeed.  So the idea is, in those cases, have a boolean expression that encodes the account existence, and use it to enable/disable the field lookups.

The way we disable a lookup is by:
1- In all tables always have an all 0 row
2- Multiply all expressions of the lookup by the "selector" expression, so that when it's zero, the lookup is a list of 0s and always suceceds, bypassing the other possibly non-zero values in the original lookup expression.

> If we add `if callee_bytecode_hash != 0`, I guess it'll panic(or just fail) if we read the balance of a non-existing account. Not sure my understanding is correct and not sure I understand your meaning correctly.

I'm not sure I understand this part

Comment (ed255) [reply src/zkevm_specs/evm/util/call_gadget.py:90]: BTW I'm working on a simple check on zkevm-circuits that will guarantee that we don't have invalid account field read/writes: https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1084

**Review by ed255 — APPROVED** [review]
LGTM! Nice work!

As I mentioned in https://github.com/privacy-scaling-explorations/zkevm-specs/pull/348#discussion_r1073778886 the logic of `if code_hash == 0, skip balance read lookup` doesn't apply in this opcode because the possibley non-existing account is the callee, whereas the balance read lookup is done on the caller.  So the PR looks good to me in the current form :)

File: src/zkevm_specs/evm/util/call_gadget.py (hunk 7 above)
Comment (ed255) [reply src/zkevm_specs/evm/util/call_gadget.py:90]: I consider this conversation resolved!  But we can continue a discussion to resolve any doubts, etc :)
